### PR TITLE
Implement ExportCommand Key Mutation Callback

### LINKED ISSUE #333: ExportCommand Key Mutation Callback
### Suggestion

Introduce a new optional property of the `ExportCommand` called `keys` where users can provide a callback function `(key: string) => string`. This function will take a `configKey` as a string, apply desired mutations (like adding a prefix or suffix), and return a modified key to be used in the export result.

### Motivation

- This addition will enable functionalities such as `--prefix` or `--suffix` in the export command of the CLI, allowing for more versatile command options.
- Adapting keys on export is a common practice in application configuration management, enhancing flexibility for different environments.
- It accommodates specific requirements of some compilers or builders that necessitate particular prefixes for environment variables.

### Context

_No response_

### PR DESCRIPTION
**Pull request by RonConfigu**
## QA Report I
- [x] Test Units Passes

Closes #333

### CODE REVIEW

File: ts/packages/ts/src/commands/ExportCommand.ts [hunk 1]
```diff
@@ -1,22 +1,44 @@
...
+
+    if (!keys) {
+      return result;
+    }
+
+    return _.mapKeys(result, (current, key) => {
+      try {
+        const mutatedKey = keys(key);
+        if (!NAME(mutatedKey)) {
+          throw new ConfigError('invalid config key', `${mutatedKey} must be a valid name`);
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/ExportCommand.ts:31]: Error message alignment with `ConfigSchema` key validation.

```suggestion
          throw new ConfigError('invalid config key', `key "${mutatedKey}" mustn't contain reserved words`);
```

File: ts/packages/ts/src/commands/ExportCommand.ts [hunk 2]
```diff
@@ -1,22 +1,44 @@
...
+        }
+        return mutatedKey;
+      } catch (error) {
+        return key;
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/ExportCommand.ts:35]: Why are we returning the original key and not throwing an exception? There is no requirement for that in the issue and in terms of UX, it makes things very unclear for the user if they supply a function that doesn't pass `NAME` validation.

File: ts/packages/ts/src/commands/ExportCommand.ts [hunk 3]
```diff
@@ -1,22 +1,44 @@
...
+      return result;
+    }
+
+    return _.mapKeys(result, (current, key) => {
+      try {
+        const mutatedKey = keys(key);
+        if (!NAME(mutatedKey)) {
+          throw new ConfigError('invalid config key', `${mutatedKey} must be a valid name`);
+        }
+        return mutatedKey;
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/ExportCommand.ts:33]: Let's maintain the error scoping in usage of the commands. This also aligns with the same logic in `ConfigSchema` key validation:
```node.js
  const errorScope: [string, string][] = [
        ['ExportCommand', ''],
        ['ConfigKey', key],
      ];
      const mutatedKey = keys(key);
      if (!NAME(mutatedKey)) {
        throw new ConfigError('invalid config key', `${mutatedKey} must be a valid name`, errorScope);
      }
      return mutatedKey;
```

File: ts/packages/ts/src/commands/commands.test.ts
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:1]: I'm kind of conflicted regarding any test that breaks the signature of `keys`. These tests cover scenarios where the user wants to break our interface intentionally as Typescript already validates this for the user of the SDK. I'm leaning towards them being redundant and unnecessary for the same reason we don't check the `typeof` of the return of `keys` being a string.

File: ts/packages/ts/src/commands/commands.test.ts [hunk 4]
```diff
@@ -222,4 +223,162 @@ describe(`commands`, () => {
...
+          },
+        }),
+        configs: {
+          KEY0: 'KEY0',
+          KEY1: 'KEY1',
+        },
+      }).run();
+
+      const exportedConfigs = await new ExportCommand({ pipe: evalResult, keys: (key) => `!${key}` }).run();
+      expect(exportedConfigs).toStrictEqual({ KEY0: 'KEY0', KEY1: 'KEY1' });
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:355]: ExportCommand should throw in this scenario, this test should be adjusted

File: ts/packages/ts/src/commands/commands.test.ts [hunk 5]
```diff
@@ -222,4 +223,162 @@ describe(`commands`, () => {
...
+        },
+      }).run();
+
+      const exportedConfigs = await new ExportCommand({
+        pipe: evalResult,
+        keys: (key) => {
+          throw new Error('bad key');
+        },
+      }).run();
+      expect(exportedConfigs).toStrictEqual({ KEY0: 'KEY0', KEY1: 'KEY1' });
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:381]: Should expect an exception

File: ts/packages/ts/src/commands/ExportCommand.ts (hunk 2 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/ExportCommand.ts:35]: There is no requirement to raise exception as well.

Refactored

Comment (RonConfigu) [reply ts/packages/ts/src/commands/ExportCommand.ts:35]: Test refactored accordingly

File: ts/packages/ts/src/commands/ExportCommand.ts (hunk 3 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/ExportCommand.ts]: done

File: ts/packages/ts/src/commands/commands.test.ts (hunk 4 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/commands.test.ts:355]: done

File: ts/packages/ts/src/commands/commands.test.ts [hunk 6]
```diff
@@ -222,4 +223,162 @@ describe(`commands`, () => {
...
+          KEY1: 'KEY1',
+        },
+      }).run();
+
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-expect-error
+      const exportedConfigs = await new ExportCommand({ pipe: evalResult, keys: (key) => 5 }).run();
+      expect(exportedConfigs).toStrictEqual({ KEY0: 'KEY0', KEY1: 'KEY1' });
+    });
+    test('Export with bad keys mutation callback - returns empty string', async () => {
```
[the remark this answers is not in the document]
Comment (RonConfigu) [reply ts/packages/ts/src/commands/commands.test.ts:315]: so? what if NAME will break or change?

File: ts/packages/ts/src/commands/commands.test.ts (hunk 5 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/commands.test.ts:381]: done

File: ts/packages/ts/src/commands/commands.test.ts
Comment (RonConfigu) [reply ts/packages/ts/src/commands/commands.test.ts:1]: better safe than sorry in my opinion.
If you want i'll delete these tests

**Comment by RonConfigu** [comment]
@RichardAkman Please review the code again. changes have been made to the command not only the tests.

File: ts/packages/ts/src/commands/commands.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (RichardAkman) [reply ts/packages/ts/src/commands/commands.test.ts:324]: Fair enough, extra tests don't hurt.

File: ts/packages/ts/src/commands/commands.test.ts
Comment (RichardAkman) [reply ts/packages/ts/src/commands/commands.test.ts:1]: Fair enough, extra tests don't hurt.

File: ts/packages/ts/src/commands/commands.test.ts [hunk 7]
```diff
@@ -222,4 +224,179 @@ describe(`commands`, () => {
...
+        configs: {
+          KEY0: 'KEY0',
+          KEY1: 'KEY1',
+        },
+      }).run();
+      await expect(
+        new ExportCommand({
+          pipe: evalResult,
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-expect-error
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:292]: NIT: Cleans up the disable next line. Check if it works for you.
```suggestion
          // @ts-expect-error - should throw ConfigError
```

File: ts/packages/ts/src/commands/commands.test.ts [hunk 8]
```diff
@@ -222,4 +224,179 @@ describe(`commands`, () => {
...
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:319]: NIT: Cleans up the disable next line. Check if it works for you.
```suggestion
          // @ts-expect-error - should throw ConfigError
```

**Comment by RonConfigu** [comment]
@RonConfigu Fix test to use method for evals

File: ts/packages/ts/src/commands/ExportCommand.ts [hunk 9]
```diff
@@ -1,22 +1,48 @@
...
+      } catch (error) {
+        throw new ConfigError('invalid config key', error.message, errorScope);
+      }
+    });
+  }
+
   async run() {
     const { pipe } = this.parameters;
-    return _.mapValues(pipe, (current) => current.result.value);
+    return _.mapValues(this.mutateKeys(pipe), (current) => current.result.value);
```
Comment (rannn505) [inline ts/packages/ts/src/commands/ExportCommand.ts:46]: For better debugging experience

```suggestion
    const configDict = _.mapValues(pipe, (current) => current.result.value);
    const keyMutatedConfigDict = this.mutateKeys(configDict);
    return keyMutatedConfigDict;
```

File: ts/packages/ts/src/commands/commands.test.ts [hunk 10]
```diff
@@ -222,4 +224,172 @@ describe(`commands`, () => {
       }
     });
   });
+  describe(`ExportCommand`, () => {
```
Comment (rannn505) [inline ts/packages/ts/src/commands/commands.test.ts:227]: Looks great!

File: ts/packages/ts/src/commands/commands.test.ts [hunk 11]
```diff
@@ -222,4 +224,172 @@ describe(`commands`, () => {
...
+          KEY0: 'KEY0',
+          KEY1: 'KEY1',
+        },
+      }).run();
+      await expect(
+        new ExportCommand({
+          pipe: evalResult,
+          keys: (key) => '',
+        }).run(),
+      ).rejects.toBeInstanceOf(ConfigError);
```
Comment (rannn505) [inline ts/packages/ts/src/commands/commands.test.ts:342]: not sure is it returning an empty string or throwing an error?
`''` is not in the NAME `RESERVED_NAMES` list

**Review by rannn505 — CHANGES_REQUESTED** [review]
@RonConfigu thats gret, take a look at my comments 🙏🏽

File: ts/packages/ts/src/commands/ExportCommand.ts [hunk 12]
```diff
@@ -1,22 +1,57 @@
...
+        ['ExportCommand', ''],
+        ['ConfigKey', key],
+      ];
+      let mutatedKey = '';
+      let isKeyValid = true;
+      try {
+        mutatedKey = String(keys(key));
+      } catch (error) {
+        isKeyValid = false;
+      }
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/ExportCommand.ts:38]: I'm not sure why we are masking errors that come from `keys`, unless there's a reason I'm not aware of, I think we should let the user see the error that comes from their own function

File: ts/packages/ts/src/commands/commands.test.ts [hunk 13]
```diff
@@ -222,4 +224,81 @@ describe(`commands`, () => {
...
+    });
+    test('Export with bad keys mutation callback - returns number', async () => {
+      const evalResult = await getEvalResult();
+      // @ts-expect-error - should throw ConfigError
+      const exportedConfigs = await new ExportCommand({ pipe: evalResult, keys: (key) => 5 }).run();
+      expect(exportedConfigs).toStrictEqual({ '5': 'KEY1' });
+    });
+    test('Export with bad keys mutation callback - returns empty string', async () => {
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:274]: `returns empty string` isn't true, it throws, please adjust the test name 🙏🏻

File: ts/packages/ts/src/commands/ExportCommand.ts (hunk 12 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/ExportCommand.ts:38]: done

File: ts/packages/ts/src/commands/commands.test.ts (hunk 13 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/commands.test.ts]: done

**Review by RichardAkman — APPROVED** [review]
LGTM 💯

**Review by rannn505 — APPROVED** [review]
@RonConfigu thanks, I wil go on to work on the CLI issue now

**Pull request merged**